### PR TITLE
[WIP] Variable Length CuDNN LSTMs/RNNs

### PR DESCRIPTION
**Pull request by bstriner**
Hi! Started working on implementing variable sequence lengths in cudnn. Getting close, but also need to implement Pack/Unpack operations to make things useful. Hopefully will have this working within a week or so. Should be a huge performance difference.

If anyone else is interested in the subject and wants to help, please let me know!

Cheers

### CODE REVIEW

**Comment by qlzh727** [comment]
Hi @bstriner, from the change description, it seems that the change is still on-going (unfinished). Can you please confirm whether it is the case?

**Comment by bstriner** [comment]
@qlzh727 Correct. Hoping to finish this soon but just started over the weekend. Getting faster as I'm getting used to the TF APIs. Figured I would put the PR out there in case anyone wanted to help or start commenting on the code. Just finished getting the kernel written for alignment. Basically, 4 kernels:

* Calculate alignment of packed sequence (done and tested)
* Pack a sequence (unwritten)
* Unpack a sequence (unwritten)
* Variable length LSTM (mostly done)

**Comment by bstriner** [comment]
Packing operations seem to be calculating everything correctly. Would appreciate any tips on making the kernels faster, maybe adding in some shared memory. Here is how they work currently.
* Pack extracts just the valid data from a sequence
* Unpack pads that data back into the original sequence shape. Padding regions are zero.

```python
import numpy as np
import tensorflow as tf
from tensorflow.contrib.cudnn_rnn.python.ops import packing_ops

sequence = tf.constant(np.arange(6 * 3).reshape((6, 3, 1)), dtype=tf.float32) + 1
sequence_lengths = tf.constant(np.array([6, 3, 2]), dtype=tf.int32)
alignments, batch_sizes = packing_ops.packed_sequence_alignment(sequence_lengths, name='align')
packed = packing_ops.pack_sequence(sequence, alignments, batch_sizes)
unpacked = packing_ops.unpack_sequence(packed, alignments, batch_sizes)

with tf.Session() as sess:
    _sequence_lengths, _sequence, _packed, _unpacked = sess.run([
        sequence_lengths, sequence, packed, unpacked
    ])
    print("Sequence lengths: {}".format(_sequence_lengths))
    print("Sequence: \n{}".format(_sequence[:, :, 0]))
    print("Packed: {}".format(_packed[:, 0]))
    print("Unpacked: \n{}".format(_unpacked[:, :, 0]))
```

> Sequence lengths: [6 3 2]
> Sequence: 
> [[ 1.  2.  3.]
>  [ 4.  5.  6.]
>  [ 7.  8.  9.]
>  [10. 11. 12.]
>  [13. 14. 15.]
>  [16. 17. 18.]]
> Packed: [ 1.  2.  3.  4.  5.  6.  7.  8. 10. 13. 16.]
> Unpacked: 
> [[ 1.  2.  3.]
>  [ 4.  5.  6.]
>  [ 7.  8.  0.]
>  [10.  0.  0.]
>  [13.  0.  0.]
>  [16.  0.  0.]]

**Comment by bstriner** [comment]
@qlzh727 Just got everything working! Could use some polish if anyone has time to give it a review, but the code seems to be working correctly and is enough to get started training. Remaining tasks (can be follow-up issues):
* Maybe refactor the packing kernels. Could probably make them faster. Lots of ways to write that kernel.
* Only made the variable length rnn for int32 indices. Should probably template that out.
* I added tests but can always add more. Only wrote tests for forward, not for backward yet.
* Should setup some benchmarking
* I'm pretty sure the descriptors are being deleted correctly but I would appreciate if anyone can take a look

Anyways, here is a worked example of a bidirectional CuDNN LSTM with variable length sequences in tensorflow. You can see the variable lengths are working correctly in the output below.

```python
import tensorflow as tf
from tensorflow.contrib.cudnn_rnn.python.layers.cudnn_rnn import CudnnLSTM, CUDNN_RNN_BIDIRECTION
from tensorflow.contrib.cudnn_rnn.python.ops import packing_ops

lengths = [6,3,2]
l = lengths[0]
n = len(lengths)
m = 1
data_dtype = tf.float32
index_dtype = tf.int32

# Create standard CudnnLSTM
lstm = CudnnLSTM(
    num_layers=1, num_units=m, dtype=data_dtype, direction=CUDNN_RNN_BIDIRECTION,
    kernel_initializer=tf.initializers.constant(0.2),
    bias_initializer=tf.initializers.constant(0.2))

# Standard 3d sequence inputs
inputs = tf.ones(shape=(l, n, m), dtype=data_dtype)
# Length of each sequence in decreasing order
sequence_lengths = tf.constant(lengths, dtype=index_dtype)
# Calculate alignments
align = packing_ops.packed_sequence_alignment(sequence_lengths)
# Pack the sequence
packed_inputs = packing_ops.pack_sequence(inputs, *align)
# Run the LSTM
packed_outputs, _ = lstm(packed_inputs, sequence_lengths=sequence_lengths)
# Unpack the outputs
unpacked_outputs = packing_ops.unpack_sequence(packed_outputs, *align)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    _sequence_lengths, _packed_outputs, _unpacked_outputs = sess.run([
        sequence_lengths, packed_outputs, unpacked_outputs])
    print("Sequence lengths: {}".format(_sequence_lengths))
    print("Unpacked outputs forward: \n{}".format(_unpacked_outputs[:, :, 0]))
    print("Unpacked outputs backward: \n{}".format(_unpacked_outputs[:, :, 1]))
```

> Sequence lengths: [6 3 2]
> Unpacked outputs forward: 
> [[0.21531965 0.21531965 0.21531965]
>  [0.35149407 0.35149407 0.35149407]
>  [0.43308067 0.43308067 0.        ]
>  [0.4820674  0.         0.        ]
>  [0.5121958  0.         0.        ]
>  [0.53126556 0.         0.        ]]
> Unpacked outputs backward: 
> [[0.53126556 0.43308067 0.35149407]
>  [0.5121958  0.35149407 0.21531965]
>  [0.4820674  0.21531965 0.        ]
>  [0.43308067 0.         0.        ]
>  [0.35149407 0.         0.        ]
>  [0.21531965 0.         0.        ]]

**Comment by qlzh727** [comment]
Thanks for the contribution. Seems to be a very large PR, which will take some time on our side to digest. 

James and Guangda, can both of you take a look first? Feel free to add more reviewer for this PR.

**Comment by albertz** [comment]
As far as I remember, you also need to sort the seqs by seq length, in descending order. Are you doing this?

Btw, check out Microsoft CNTK code which wraps cuDNN which seems to do all of that correctly.

**Comment by bstriner** [comment]
@albertz yes, but that uses ops in tensorflow already. An actual example goes like this. Will definitely need some examples for people to get it right.

```python
        u_order = argsort(utterance_lengths, direction='DESCENDING')
        u_backorder = argsort(u_order)
        sorted_lengths = tf.gather(utterance_lengths, indices=u_order, axis=0)
        sorted_utterances = tf.gather(utterances, indices=u_order, axis=1)
        u_align = packed_sequence_alignment(sorted_lengths)
        u_packed = pack_sequence(sorted_utterances, *u_align)

        with tf.variable_scope("Listener") as listener_scope:
            h, _ = CudnnLSTM(num_layers=3, num_units=params.listener_dim, direction=CUDNN_RNN_BIDIRECTION)(
                u_packed, sequence_lengths=sorted_lengths
            )
            packed_logits = slim.fully_connected(h, num_outputs=vocab_size+1, activation_fn=None, scope='logits')
            logits = unpack_sequence(packed_logits, *u_align)
            logits = tf.gather(logits, indices=u_backorder, axis=1)
```

**Comment by bstriner** [comment]
Actually, it might be worth making a combined operation that does the sorting and packing together. The LSTM itself is pretty straightforward. The packing operations are where we might want to make some changes.

**Comment by bstriner** [comment]
I've just started training on a BiLSTM with CTC loss. Iterations per second goes from 0.29 to 3.15102 by switching from `bidirectional_dynamic_rnn` to `CudnnLSTM`. So maybe the packing and sorting could be faster, but I'm getting about 10x speedup already.

This is on sequence lengths about 2500, 64 dims, batch size 8, 3 layer bilstm with 256 units.

**Comment by bstriner** [comment]
I'm thinking it might actually make sense to rewrite pack and unpack as reshapes and scatters.

**Comment by bstriner** [comment]
So it would be possible to do the sorting and packing together.
* An operation with inputs (sequence lengths, batch_order) that outputs 2 alignments(fwd_align, back_align)
* packing is gather(reshape(x, (s*b, d)), fwd_align, axis=0)
* unpacking is reshape(scatter(x, back_align, axis=0, output_Shape=(s*b,d)), (s,b,d))

**Comment by bstriner** [comment]
Hi guys! Just redid the packing. Now there is just one custom operation `SequenceGatherScatterIndices`. This operation takes the total sequence length, the sorted sequence lengths, and the sort order. It produces a set of indices that can be used to pack and unpack using gather_nd and scatter_nd, so no longer need those custom ops for packing and unpacking. The single gather_nd or scatter_nd will both remove the padding and sort the batches, so you can do both at the same time.

**Comment by bstriner** [comment]
I've been training a lot over the weekend and everything seems to be working right. I'm thinking about dropping the packing kernels entirely. It seems like I can get the correct indices with existing operations by doing the below. Might be faster or slower than the custom kernel but definitely easier to write and maintain. Using the fact that boolean_mask seems to take things in deterministic row-major order.

```python
def sequence_indices(sorted_sequence_lengths, batch_order):
    mg = tf.meshgrid(tf.range(sorted_sequence_lengths[0]), batch_order, indexing='ij')
    mg = tf.stack(mg, axis=-1)
    mask = tf.transpose(tf.sequence_mask(lengths=sorted_sequence_lengths, maxlen=l), (1, 0))
    idx = tf.boolean_mask(mg, mask)
    return idx
```

The LSTM itself seems to be about right. The only improvement might be caching the descriptors somehow. Should be able to reuse the forward descriptors for the backward pass instead of regenerating but I'm not sure exactly where that would go and how much of an improvement that would actually make.

**Comment by bstriner** [comment]
@albertz @qlzh727 Just did some cleanup and reverted a fair number of no-longer-necessary changes. Only major remaining tasks I think are documentation, formatting and maybe some extra testing. You can use `sequence_gather_scatter_indices` to go from unsorted sequence lengths to indices. You can then `gather_nd` and `scatter_nd` to use that sequence with variable length `CudnnLSTM`.

Does anyone know more about `V2`? I'm not sure what tests should be added for that specifically. There is also this comment in the code:
```
# TODO(jamesqin): switch default value to "1" on May 25th 2018, and get rid
  # of V1 ops. 
```

**Comment by bstriner** [comment]
Looks like my editor trashed permissions so I'll reset those when I get a chance.

**Comment by protoget** [comment]
Thank you Ben. I think it's likely this feature would be support natively in later Cudnn versions. I'd suggest not making more changes to it for now.

**Comment by bstriner** [comment]
@protoget Why would that be the case? I wouldn't expect nvidia to drop support. You can't emulate bilstm with just masking and it is >10x faster than a `bidirectional_dynamic_rnn`.

**Comment by bstriner** [comment]
Actually, maybe I'm not clear on what you mean. You think the creating of the descriptors is going to be rolled into CuDNN? That sounds cool.

**Comment by protoget** [comment]
I mean later CudnnRNNForwardTrain() API in Cudnn might support sequence_length as a new param, so TF side we don't need to add additional ops to handle padding.

**Comment by bstriner** [comment]
Do you have a link or something? `cudnnRNNForwardTraining` has `seqLength` but that is just an int.

**Comment by bstriner** [comment]
Aha! Some insider stuff I'll let it be. The PR as is works and gets my model running in reasonable time. I'll leave it up just-in-case but sounds like there will be some other changes coming.

Would probably still need a function like `sequence_indices` to do the packing and unpacking but that would be a small change.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc [hunk 1]
```diff
@@ -1225,9 +1242,40 @@ class CudnnRnnSequenceTensorDescriptor
...
+            int strides[] = {data_size, 1, 1};
+            RETURN_IF_CUDNN_ERROR(cudnnSetTensorNdDescriptor(
+                /*tensorDesc=*/tensor_desc.get(), /*dataType=*/data_type,
+                /*nbDims=*/3, /*dimA=*/dims,
+                /*strideA=*/strides));
+        }
+        auto descriptors = std::vector<TensorDescriptor>(0);
+        auto handles = std::vector<cudnnTensorDescriptor_t>(seq_length);
+        int pos = batch_size-1;
+        for(int i=0; i< seq_length; i++){
```
Comment (shahzadlone) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: prefer pre-inc over post-inc
i.e. ```++i```

File: tensorflow/stream_executor/cuda/cuda_dnn.cc [hunk 2]
```diff
@@ -1225,9 +1242,40 @@ class CudnnRnnSequenceTensorDescriptor
...
+            while(sequence_lengths[pos] <= i && pos > 0){
+                pos--;
+            }
+            auto td = CreateTensorDescriptor();
```
Comment (shahzadlone) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: perhaps ```const``` if td has a well written move constructor (or default copy and move can be used just fine), I saw that because i see you std::move it after.

**Comment by gbaned** [comment]
@shahzadlone Any update please ?

**Comment by gbaned** [comment]
@shahzadlone gentle ping

**Comment by gbaned** [comment]
@bstriner could you please resolve the conflicts? Thanks!

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
